src/main/php/peer/server/AsyncServer.class.php

```diff
@@ -0,0 +1,217 @@
+<?php namespace peer\server;
+
+use lang\{IllegalStateException, Throwable};
+use peer\server\protocol\SocketAcceptHandler;
+use peer\{ServerSocket, SocketException, SocketTimeoutException};
+
+/**
+ * Asynchronous TCP/IP Server
+ *
+ * ```php
+ * use peer\server\AsyncServer;
+ *   
+ * $server= new AsyncServer();
+ * $server->listen(new ServerSocket('127.0.0.1', 6100), new MyProtocol());
+ * $server->service();
+ * ```
+ *
+ * @see   xp://peer.ServerSocket
+ * @test  xp://peer.unittest.server.AsyncServerTest
+ */
+class AsyncServer extends Server {
+  private $select= [], $handle= [], $tasks= [];
+
+  /**
+   * Adds server socket to listen on, associating protocol handler with it
+   *
+   * @param  peer.ServerSocket|peer.BSDServerSocket $socket
+   * @param  peer.server.ServerProtocol $protocol
+   * @return self
+   */
+  public function listen($socket, $protocol) {
+    $protocol->server= $this;
+
+    $socket->create();
+    $socket->bind(true);
+    $socket->listen();
+
+    $this->select[]= $socket;
+    $this->handle[]= [function($socket) use($protocol) {
+      $connection= $socket->accept();
+      if ($protocol instanceof SocketAcceptHandler && !$protocol->handleAccept($connection)) {
+        $connection->close();
+        return;
+      }
+
+      $this->tcpnodelay && $connection->useNoDelay();
+      $protocol->handleConnect($connection);
+
+      $this->select[]= $connection;
```

Comment:
This will never reuse any existing array indices:

```bash
$ xp -d '$a= []; $a[]= "a"; unset($a[0]); $a[]= "b"; return $a'
array(1) {
  [1]=>
  string(1) "b"
}
```

...which will at some point run into the following problem:

```bash
$ xp -d '$a= [PHP_INT_MAX => "a"]; $a[]= "b"; return $a;'
Exception lang.Error (Cannot add element to the array as the next element is already occupied)
```

Reply:
Solved by reusing array indices